taskchampion/taskchampion/src/server/cloud/server.rs

```diff
@@ -0,0 +1,1130 @@
+            self.cleanup()
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Perform cleanup, deleting unnecessary data.
+    fn cleanup(&mut self) -> Result<()> {
+        // Construct a vector containing all (child, parent, creation) tuples
+        let mut versions = self
+            .service
+            .list(b"v-")
+            .filter_map(|res| match res {
+                Ok(ObjectInfo { name, creation }) => {
+                    if let Some((p, c)) = Self::parse_version_name(&name) {
+                        Some(Ok((c, p, creation)))
+                    } else {
+                        None
+                    }
+                }
+                Err(e) => Some(Err(e)),
+            })
+            .collect::<Result<Vec<_>>>()?;
+        versions.sort();
+
+        // Function to find the parent of a given child version in `versions`, taking
+        // advantage of having sorted the vector by child version ID.
+        let parent_of = |c| match versions.binary_search_by_key(&c, |tup| tup.0) {
+            Ok(idx) => Some(versions[idx].1),
+            Err(_) => None,
+        };
+
+        // Create chains mapping forward (parent -> child) and backward (child -> parent), starting
+        // at "latest".
+        let mut fwd_chain = HashMap::new();
+        let mut rev_chain = HashMap::new();
+        let mut iterations = versions.len() + 1; // For cycle detection.
+        let latest = self.get_latest()?;
+        if let Some(mut c) = latest {
+            while let Some(p) = parent_of(c) {
+                fwd_chain.insert(p, c);
+                rev_chain.insert(c, p);
+                c = p;
+                iterations -= 1;
+                if iterations == 0 {
+                    return Err(Error::Server("Version cycle detected".into()));
+                }
+            }
+        }
+
+        // Collect all versions older than MAX_VERSION_AGE_SECS
+        #[cfg(not(test))]
+        let age_threshold = {
+            let now = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .map(|t| t.as_secs())
+                .unwrap_or(0);
+            now.saturating_sub(MAX_VERSION_AGE_SECS)
+        };
+
+        // In testing, cutoff age is 1000.
+        #[cfg(test)]
+        let age_threshold = 1000;
+
+        let old_versions: HashSet<Uuid> = versions
+            .iter()
+            .filter_map(|(c, _, creation)| {
+                if *creation < age_threshold {
+                    Some(*c)
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        // Now, any pair not present in that chain can be deleted. However, another replica
+        // may be in the state where it has uploaded a version but not changed "latest" yet,
+        // so any pair with parent equal to latest is allowed to stay.
+        for (c, p, _) in versions {
+            if fwd_chain.get(&p) != Some(&c) && Some(p) != latest {
```

Comment:
Seems like we could eliminate the need for `fwd_chain`, which doesn't appear to be used elsewhere:

```suggestion
            if rev_chain.get(&c) != Some(&p) && Some(p) != latest {
```

Reply:
It is used!

Reply:
Hmm? This is the only use I see.

Reply:
Oh, you're right! I thought this comment was on the `fwd_chain.insert` and saw `fwd_chain.get` later. Yes, the `rev_chain` will do just as well here. Thanks!